File: raffia-core/src/main/java/com/codepoetics/raffia/api/PropertySet.java
package com.codepoetics.raffia.api;

import org.pcollections.HashTreePMap;
import org.pcollections.PMap;
import org.pcollections.PVector;
import org.pcollections.TreePVector;

import java.util.*;

public class PropertySet implements Iterable<ObjectEntry> {

  public static PropertySet of(ObjectEntry...entries) {
    return of(Arrays.asList(entries));
  }

  public static PropertySet of(Collection<ObjectEntry> entries) {
    Map<String, Basket> properties = new LinkedHashMap<>();
    for (ObjectEntry entry : entries) {
      properties.put(entry.getKey(), entry.getValue());
    }
    return of(properties);
  }

  public static PropertySet of(Map<String, Basket> properties) {
    return new PropertySet(TreePVector.from(properties.keySet()), HashTreePMap.from(properties));
  }

  private final PVector<String> keys;
  private final PMap<String, Basket> properties;

  private PropertySet(PVector<String> keys, PMap<String, Basket> properties) {
    this.keys = keys;
    this.properties = properties;
  }

  public PropertySet with(String key, Basket basket) {
    return properties.containsKey(key)
      ? new PropertySet(keys, properties.plus(key, basket))
      : new PropertySet(keys.plus(key), properties.plus(key, basket));
  }

  public int size() {
    return properties.size();
  }

  public Basket get(String key) {
    return properties.get(key);
  }

  public boolean containsKey(String key) {
    return properties.containsKey(key);
  }

  public PropertySet minus(String key) {
    return new PropertySet(keys.minus(key), properties.minus(key));
  }

  public Map<String, Basket> toMap() {
    Map<String, Basket> result = new LinkedHashMap<>();
    for (String key : keys) {
      result.put(key, properties.get(key));
    }
    return result;
  }

  @Override
  public boolean equals(Object other) {
    return this == other
        || (other instanceof PropertySet
          && PropertySet.class.cast(other).keys.equals(keys)
          && PropertySet.class.cast(other).properties.equals(properties));
  }

  @Override
  public int hashCode() {
    return Objects.hash(keys.hashCode(), properties.hashCode());
  }

  @Override
  public String toString() {
    return toMap().toString();
  }

  @Override
  public Iterator<ObjectEntry> iterator() {
    final Iterator<String> keyIterator = keys.iterator();
    return new Iterator<ObjectEntry>() {
      @Override
      public void remove() {
        throw new UnsupportedOperationException();
      }

      @Override
      public boolean hasNext() {
        return keyIterator.hasNext();
      }

      @Override
      public ObjectEntry next() {
        String key = keyIterator.next();
        return ObjectEntry.of(key, properties.get(key));
      }
    };
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/baskets/PropertySet.kt
package com.codepoetics.raffia.baskets

import com.codepoetics.raffia.java.api.Mapper
import org.pcollections.HashTreePMap
import org.pcollections.PMap
import org.pcollections.PVector
import org.pcollections.TreePVector

import java.util.*

data class ObjectEntry(val key: String, val value: Basket) {
    override fun toString(): String {
        return key + ": " + value
    }

    companion object {
        @JvmStatic
        fun of(key: String, value: Basket): ObjectEntry {
            return ObjectEntry(key, value)
        }
    }
}

class PropertySet private constructor(private val keys: PVector<String>, private val properties: PMap<String, Basket>)
    : Iterable<ObjectEntry> {

    fun with(key: String, basket: Basket): PropertySet =
        if (properties.containsKey(key))
            PropertySet(keys, properties.plus(key, basket))
        else
            PropertySet(keys.plus(key), properties.plus(key, basket))


    fun size(): Int = properties.size

    operator fun get(key: String): Basket? = properties[key]

    fun containsKey(key: String): Boolean = properties.containsKey(key)

    operator fun minus(key: String): PropertySet = PropertySet(keys.minus(key), properties.minus(key))

    fun toMap(): Map<String, Basket> {
        val result = LinkedHashMap<String, Basket>()
        keys.forEach { result.put(it, properties[it]!!) }
        return result
    }

    val isEmpty: Boolean
        get() = properties.isEmpty()

    override fun equals(other: Any?): Boolean {
        return this === other || other is PropertySet
                && PropertySet::class.java.cast(other).keys == keys
                && PropertySet::class.java.cast(other).properties == properties
    }

    override fun hashCode(): Int {
        return Objects.hash(keys.hashCode(), properties.hashCode())
    }

    override fun toString(): String {
        return toMap().toString()
    }

    override fun iterator(): Iterator<ObjectEntry> {
        val keyIterator = keys.iterator()
        return object : Iterator<ObjectEntry> {
            override fun hasNext(): Boolean = keyIterator.hasNext()

            override fun next(): ObjectEntry = keyIterator.next().let { ObjectEntry.of(it, properties[it]!!) }
        }
    }

    fun mapValues(valueMapper: Mapper<Basket, Basket>): PropertySet =
        PropertySet.of(map { entry -> ObjectEntry.of(entry.key, valueMapper.map(entry.value)) })

    fun mapEntries(entryMapper: Mapper<ObjectEntry, List<ObjectEntry>>): PropertySet =
            PropertySet.of(flatMap { entry -> entryMapper.map(entry) })

    companion object {

        @JvmStatic
        fun of(vararg entries: ObjectEntry): PropertySet {
            return of(Arrays.asList(*entries))
        }

        @JvmStatic
        fun of(entries: Collection<ObjectEntry>): PropertySet {
            val properties = LinkedHashMap<String, Basket>()
            for (entry in entries) {
                properties.put(entry.key, entry.value)
            }
            return of(properties)
        }

        @JvmStatic
        fun of(properties: Map<String, Basket>): PropertySet {
            return PropertySet(TreePVector.from(properties.keys), HashTreePMap.from(properties))
        }
    }
}
